### PR TITLE
ENH: add cpp studentized range implementations

### PR DESCRIPTION
**Pull request by DominicChm**
#### What does this implement/fix?
Test PR to implement a CPP version of the studentized range. 

(Sorry if this somehow ends up in the main scipy repo)

### CODE REVIEW

**Comment by DominicChm** [comment]
Apperently git ignores the .so (compiled) files, so compile the thing by running `gcc -shared -fPIC -o studentized_range.so studentized_range.cpp` in the studentized_range directory

**Comment by swallan** [comment]
Thanks! -- I will be looking soon. Have to set up a Linux VM as I foolishly updated to a new version of MacOS that doesn't have gcc support.

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -3861,8 +3861,7 @@ def tukeykramer(*args, sig_level=.05):
 
     def get_q(p, r, v, guess=3):
         def wrapper_cdf(q_crit, treatments, ddof, alpha):
-            return alpha - (1 - prtrng(q_crit, ddof,
-                                       treatments)[0])
+            return alpha - (1 - studentized_cdf(q_crit, treatments, ddof))
```
Comment (swallan) [inline scipy/stats/stats.py:3864]:
```suggestion
            return alpha - (1 - studentized_cdf(q_crit, treatments, ddof)) 
```

**Comment by swallan** [comment]
I don't think I have permissions to push to the PR from my remote, but you could add 
``` 
from ._studentized_range import studentized_cdf
```
to `scipy/scipy/stats/__init__.py` so that it's accessible at `scipy.stats.studentized_cdf`

**Comment by swallan** [comment]
I think this looks good. From what I can tell, usually very close to the output of the Fortran code we were thinking about. I graphed the differences between the CDFs:
image
## this is incorrectly titled. It should be inverse cdf.
As well as the difference between the PDFs (wrapped CDF with scipy.optimize.root).
image
An undetermined reason causes my wrappers to fail at anything past 40 treatment groups, I am still trying to nail down the reason for this. 


From what I have seen, the order of the differences is in the thousands which is definitely acceptable given that there aren't many reputable existing calculations.

The C version seems to be about 3 orders of magnitudes slower, which is unfortunate (.00576s v.s. .0000091s):
Screen Shot 2020-11-23 at 9 33 13 PM
 


Similar story for the wrapped version: (.049s vs .000225s, 2 orders of magnitude)
Screen Shot 2020-11-23 at 9 36 34 PM

File: scipy/stats/studentizedrange/studentized_range.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+//PDF approximation
+double Phi(double value) {
+   return 0.5 * erfc(-value * M_SQRT1_2);
+}
+
+//PDF approximation
+double phi(double z)
+{
+    static const double inv_sqrt_2pi = 0.3989422804014327;
+    return inv_sqrt_2pi * std::exp(-0.5f * z * z);
```
Comment (swallan) [inline scipy/stats/studentizedrange/studentized_range.cpp:19]:
```suggestion
    return inv_sqrt_2pi * ::exp(-0.5f * z * z);
```
MacOS doesn't seem to like trying to get `exp` from `std` namespace:
```
studentized_range.cpp:19:32: error: no member named 'exp' in namespace 'std'
    return inv_sqrt_2pi * std::exp(-0.5f * z * z);
```

**Comment by DominicChm** [comment]
> An undetermined reason causes my wrappers to fail at anything past 40 treatment groups, I am still trying to nail down the reason for this.

Past k=40, an asymptomatic version is used rather than the standard version, which uses only one integral rather than two, which I just found an error in. Was integrating from 0 to inf, rather than -inf to inf. Oops...

> MacOS doesn't seem to like trying to get exp from std namespace:

I have the wrong import. Had `#include <math.h>` when[ apparently the correct one is](https://stackoverflow.com/questions/16518210/sqrt-is-not-a-member-of-std) `#include <cmath>`. Not sure why it compiled for me in the first place, but I guess that's just part of the C++ mystique.

> The C version seems to be about 3 orders of magnitudes slower, which is unfortunate (.00576s v.s. .0000091s):

That could definitely be a problem, especially with a large number of groups. It might be possible to optimize the double integral, which is where I assume most of the overhead is.

**Comment by swallan** [comment]
Thanks for the fix -- it is now working correctly. Interestingly, we can see exactly where the swap to the asymptotic version is used in the difference compared to the fortran in use of the wrapper. The CDF's graph is near identical to the prior version PDF (wrapper) is different:

## these are mislabelled. It should be inverse cdf.
image


On a log scale so you can see magnitude of difference:
image

**Comment by DominicChm** [comment]
Looks like we might want to delay the switch then to more closely match it. Right now it's at k=40, but it looks like the fortran version switches at 120.

**Comment by swallan** [comment]
> Looks like we might want to delay the switch then to more closely match it. Right now it's at k=40, but it looks like the fortran version switches at 120.

Something else we might want to consider is comparing both to other published sources. When the difference is < 10^-3 any differences would be hard to see, but there might be some areas that could be relevant.

**Comment by swallan** [comment]
@DominicChm - you may be interested in this: https://github.com/swallan/times_accuracy

This is your method compared with 
- Fortran
- statsmodels interpolation
- mpmath tanh-sinh method
- dblquad 

Takeaways: Assuming that MPMATH @ 19 DPS is ground truth, this implememtation is often more accurate than the fortran implementation, albeit with some failures in some cases. It is seemingly middle road in terms of speed, slower than fortran but much faster than mpmath.

**Pull request closed without merging**